File: src/main/java/com/shockang/study/algorithm/java/leetcode/leetcode201_300/leetcode205/solution2/Solution.java
package com.shockang.study.algorithm.java.leetcode.leetcode201_300.leetcode205.solution2;

/**
 * 2 个 256 数组
 *
 * @author Shockang
 */
public class Solution {
	public boolean isIsomorphic(String s, String t) {
		char[] chars = s.toCharArray();
		char[] chart = t.toCharArray();
		int[] preIndexOfs = new int[256];
		int[] preIndexOft = new int[256];
		for (int i = 0; i < chars.length; i++) {
			if (preIndexOfs[chars[i]] != preIndexOft[chart[i]]) {
				return false;
			}
			preIndexOfs[chars[i]] = i + 1;
			preIndexOft[chart[i]] = i + 1;
		}
		return true;
	}
}

File: src/main/scala/com/shockang/study/algorithm/scala/leetcode/leetcode201_300/leetcode205/solution2/Solution.scala
package com.shockang.study.algorithm.scala.leetcode.leetcode201_300.leetcode205.solution2

import com.shockang.study.algorithm.scala.leetcode.common.ListNode

/**
 * 递归
 *
 * @author Shockang
 */
object Solution {
  def reverseList(head: ListNode): ListNode = {
    if (head == null || head.next == null) return head
    val next = head.next
    val newHead = reverseList(next)
    next.next = head
    head.next = null
    newHead
  }
}
